dev/MRTCore/mrt/Core/src/MRM.cpp

```diff
@@ -899,26 +906,53 @@ STDAPI MrmGetFilePathFromName(_In_ PCWSTR filename, _Outptr_ PWSTR* filePath)
 
     pointerToPath = pathAllocated ? pathAllocated.get() : path;
 
-    size_t lengthInSizeT;
-    RETURN_IF_FAILED(StringCchLengthW(pointerToPath, STRSAFE_MAX_CCH, &lengthInSizeT));
-    length = static_cast<DWORD>(lengthInSizeT);
+    std::unique_ptr<wchar_t, decltype(&MrmFreeResource)> finalPath(nullptr, MrmFreeResource);
 
-    RETURN_IF_FAILED(StringCchLengthW(filename, STRSAFE_MAX_CCH, &lengthInSizeT));
-    DWORD lengthOfName = static_cast<DWORD>(lengthInSizeT);
+    // Will build the path at most twice.
+    // First time using current path. If not exist, do another time with parent path.
+    for (int i = 0; i < 2; i++)
+    {
+        size_t lengthInSizeT;
+        RETURN_IF_FAILED(StringCchLengthW(pointerToPath, STRSAFE_MAX_CCH, &lengthInSizeT));
+        length = static_cast<DWORD>(lengthInSizeT);
+
+        RETURN_IF_FAILED(StringCchLengthW(filename, STRSAFE_MAX_CCH, &lengthInSizeT));
+        DWORD lengthOfName = static_cast<DWORD>(lengthInSizeT);
+
+        RETURN_IF_FAILED(DWordAdd(length, lengthOfName, &length));
+        RETURN_IF_FAILED(DWordAdd(length, 1, &length));
+        RETURN_IF_FAILED(DWordMult(length, sizeof(wchar_t), &size));
+
+        PWSTR rawOutputPath = reinterpret_cast<PWSTR>(MrmAllocateBuffer(size));
+        RETURN_IF_NULL_ALLOC(rawOutputPath);
 
-    RETURN_IF_FAILED(DWordAdd(length, lengthOfName, &length));
-    RETURN_IF_FAILED(DWordAdd(length, 1, &length));
-    RETURN_IF_FAILED(DWordMult(length, sizeof(wchar_t), &size));
+        std::unique_ptr<wchar_t, decltype(&MrmFreeResource)> outputPath(rawOutputPath, MrmFreeResource);
+
+        RETURN_IF_FAILED(StringCchCopyW(outputPath.get(), length, pointerToPath));
+        RETURN_IF_FAILED(StringCchCatW(outputPath.get(), length, filename));
+
+        if (GetFileAttributes(outputPath.get()) != INVALID_FILE_ATTRIBUTES)
+        {
+            // The file exists. Done.
+            finalPath.swap(outputPath);
+            break;
+        }
 
-    PWSTR rawOutputPath = reinterpret_cast<PWSTR>(MrmAllocateBuffer(size));
-    RETURN_IF_NULL_ALLOC(rawOutputPath);
+        if (i == 0)
+        {
+            // If none of the file exists, will return the file in current path
```

Comment:
>will return the file in current path [](start = 43, length = 36)

But if the file exists in the current path, we won't even get here. We will break above.

Reply:
this is when the file doesn't exist in either location, we will return the file path in current folder.

Reply:
but if we're here, the file doesn't exist in the current folder, no?

---
In reply to: [591996835](https://github.com/microsoft/ProjectReunion/pull/554#discussion_r591996835) [](ancestors = 591996835)

Reply:
correct. this function returns a filename. If the file doesn't exist, it will still return you a name. There is no point to fail.